File: fetch-stats.js
#!/usr/bin/env node

/**
 * Fetches Cooper Flagg's career stats from balldontlie.io and writes data/stats.json.
 *
 * Usage: BALLDONTLIE_API_KEY=your_key node fetch-stats.js
 */

const fs = require('fs');
const path = require('path');

const LEBRON_TOTAL = 40474;
const FLAGG_BIRTH_YEAR = 2006;
const API_BASE = 'https://api.balldontlie.io/nba/v1';

async function apiFetch(path, apiKey) {
  const res = await fetch(`${API_BASE}${path}`, {
    headers: { Authorization: apiKey },
  });
  if (!res.ok) {
    console.error(`API ${path} responded ${res.status}: ${await res.text()}`);
    process.exit(1);
  }
  return res.json();
}

async function fetchStats() {
  const apiKey = process.env.BALLDONTLIE_API_KEY;
  if (!apiKey) {
    console.error('Missing BALLDONTLIE_API_KEY environment variable.');
    process.exit(1);
  }

  // Look up Flagg's player ID by name
  const players = await apiFetch('/players?search=Cooper+Flagg', apiKey);
  if (!players.data || players.data.length === 0) {
    console.error('Could not find Cooper Flagg in player database.');
    process.exit(1);
  }
  const playerId = players.data[0].id;
  console.log(`Found Cooper Flagg: player ID ${playerId}`);

  // Fetch season averages for current season
  const currentSeason = new Date().getFullYear();
  const body = await apiFetch(
    `/season_averages?season=${currentSeason}&player_id=${playerId}`,
    apiKey
  );

  if (!body.data || body.data.length === 0) {
    console.error('No season data returned for Flagg.');
    process.exit(1);
  }

  let totalPoints = 0;
  let totalGames = 0;

  for (const season of body.data) {
    const games = season.games_played || 0;
    const pts = season.pts || 0;
    totalPoints += Math.round(pts * games);
    totalGames += games;
  }

  const ppg = totalGames > 0 ? Math.round((totalPoints / totalGames) * 10) / 10 : 0;
  const percentage = Math.round((totalPoints / LEBRON_TOTAL) * 1000) / 10;
  const currentYear = new Date().getFullYear();

  let projectedYear = null;
  let projectedAge = null;

  if (ppg > 0) {
    const seasonsRemaining = (LEBRON_TOTAL - totalPoints) / (ppg * 82);
    projectedYear = currentYear + Math.ceil(seasonsRemaining);
    projectedAge = projectedYear - FLAGG_BIRTH_YEAR;
  }

  const data = {
    points: totalPoints,
    games: totalGames,
    ppg,
    percentage,
    projected_year: projectedYear,
    projected_age: projectedAge,
    lebron_total: LEBRON_TOTAL,
    updated: new Date().toISOString(),
  };

  const outDir = path.join(__dirname, 'data');
  if (!fs.existsSync(outDir)) {
    fs.mkdirSync(outDir);
  }

  fs.writeFileSync(path.join(outDir, 'stats.json'), JSON.stringify(data, null, 2) + '\n');
  console.log('Wrote data/stats.json:', data);
}

fetchStats();

File: fetch-stats.py
#!/usr/bin/env python3
"""Fetch Cooper Flagg's career stats and write data/stats.json.

Tries nba_api first; if NBA.com blocks the request (common in CI),
falls back to the existing stats.json so the workflow doesn't fail.
Run locally to get fresh data, then CI will commit the update.
"""

import json
import math
import os
import sys
import time
from datetime import datetime

LEBRON_TOTAL = 40474
FLAGG_BIRTH_YEAR = 2006
FLAGG_PLAYER_ID = 1642843

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
OUT_PATH = os.path.join(SCRIPT_DIR, "data", "stats.json")


def fetch_from_nba():
    from nba_api.stats.endpoints import playercareerstats

    for attempt in range(3):
        try:
            career = playercareerstats.PlayerCareerStats(
                player_id=FLAGG_PLAYER_ID, timeout=60
            )
            df = career.get_data_frames()[0]
            return int(df["PTS"].sum()), int(df["GP"].sum())
        except Exception as e:
            print(f"Attempt {attempt + 1} failed: {e}")
            if attempt < 2:
                time.sleep(5 * (attempt + 1))
    return None, None


def build_data(total_points, total_games):
    ppg = round(total_points / total_games, 1) if total_games > 0 else 0
    percentage = round((total_points / LEBRON_TOTAL) * 100, 1)
    current_year = datetime.now().year

    if ppg > 0:
        seasons_remaining = (LEBRON_TOTAL - total_points) / (ppg * 82)
        projected_year = current_year + math.ceil(seasons_remaining)
        projected_age = projected_year - FLAGG_BIRTH_YEAR
    else:
        projected_year = None
        projected_age = None

    return {
        "points": total_points,
        "games": total_games,
        "ppg": ppg,
        "percentage": percentage,
        "projected_year": projected_year,
        "projected_age": projected_age,
        "lebron_total": LEBRON_TOTAL,
        "updated": datetime.utcnow().isoformat() + "Z",
    }


def main():
    total_points, total_games = fetch_from_nba()

    if total_points is None:
        print("NBA.com fetch failed. Keeping existing stats.json.")
        if os.path.exists(OUT_PATH):
            sys.exit(0)
        else:
            sys.exit("No existing stats.json and fetch failed.")

    data = build_data(total_points, total_games)

    os.makedirs(os.path.dirname(OUT_PATH), exist_ok=True)
    with open(OUT_PATH, "w") as f:
        json.dump(data, f, indent=2)
        f.write("\n")

    print(f"Wrote {OUT_PATH}: {data}")


if __name__ == "__main__":
    main()
